draft-barnes-acme.md

```diff
@@ -218,304 +219,275 @@ In this section, we describe the four certificate management functions that ACME
+* "identifier" the identifier submitted by the client.
+* "key": the key used to verify the client's JWS request (i.e., the contents of the "jwk" field in the JWS header)
+* "status": SHOULD be "pending" (MAY be omitted)
+* "challenges" and "combinations": As selected by the server's policy for this identifier
+* "contact": the "contact" field submitted by the client, if provided
+* The "expires" and "recoveryToken" fields MUST be absent.
 
-type (required, string):
-: "challenge"
+The server allocates a new URI for this authorization, and returns a 201 (Created) response, with the authorization URI in a Location header field, and the JSON authorization object in the body.
 
-sessionID (required, string):
-: An opaque string that allows the server to correlate transactions related to this challenge request.
+~~~~~~~~~~
 
-nonce (required, string):
-: A base64-encoded octet string that the client is expected to sign with the private key of the key pair being authorized.
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: https://example.com/authz/asdf
 
-challenges (required, array):
-: A list of challenges to be fulfilled by the client in order to prove possession of the identifier.  The syntax for challenges is described in Section {{identifier-validation-challenges}}.
+{
+  "identifier": {
+    "type": "domain",
+    "value": "example.org"
+  },
 
-combinations (optional, array of arrays):
-: A collection of sets of challenges, each of which would be sufficient to prove possession of the identifier.  Clients SHOULD complete a set of challenges that that covers at least one set in this array.  Challenges are represented by their associated zero-based index in the challenges array.
+  "key": { /* JWK from JWS header */ },
 
-For example, if the server wants to have the client demonstrate both that the client controls the domain name in question, and that this client is the same client that previously requested authorization for the domain name, it might issue the following request.  The client is expected to provide "simpleHttps" and "recoveryToken" responses ("[0,2]"), or else "dns" and "recoveryToken" responses ("[1,2]"), or all three.
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
+  ],
 
-~~~~~~~~~~
-{
-  "type": "challenge",
-  "sessionID": "aefoGaavieG9Wihuk2aufai3aeZ5EeW4",
-  "nonce": "czpsrF0KMH6dgajig3TGHw",
-  "challenges": [
-    {
-      "type": "simpleHttps",
+  "challenges": {
+    "simpleHttps": {
```

Comment:
This is a good point. For instance, in the future we will want to issue wildcard certificates. You can imagine wanting to do two different SimpleHTTPS challenges to two different random subdomains, to verify full control of the domain.

Both this and the previous list/indices approach also introduce an awkward non-locality in evaluating the data structure. Proposal: Snap the pointers. Instead of `challenges`, and `combinations`, have a single field `challengeSets`, which is an array of arrays of challenges. This may mean a small amount of duplication, but I think it simplifies the structure.